element-connector/src/main/java/org/eclipse/californium/elements/tcp/NettyContextUtils.java

```diff
@@ -0,0 +1,178 @@
+			if (null != sslSession) {
+				try {
+					Principal principal = sslSession.getPeerPrincipal();
+					if (null == principal) {
+						LOGGER.log(Level.WARNING, "Principal missing");
+					} else {
+						LOGGER.log(Level.FINER, "Principal {0}", principal.getName());
+					}
+					return principal;
+				} catch (SSLPeerUnverifiedException e) {
+					LOGGER.log(Level.WARNING, "Principal {0}", e.getMessage());
+					/* ignore it */
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Build correlation context related to the provided channel.
+	 * 
+	 * @param channel channel of correlation context
+	 * @return correlation context, or null, if yet not available.
+	 */
+	public static CorrelationContext buildCorrelationContext(Channel channel) {
+		String id = channel.id().asShortText();
+		SslHandler handler = getSslHandler(channel);
+		if (null != handler) {
+			SSLEngine sslEngine = handler.engine();
+			SSLSession sslSession = sslEngine.getSession();
+			if (null != sslSession) {
+				byte[] sessionId = sslSession.getId();
+				if (null != sessionId && 0 < sessionId.length) {
+					String sslId = toHexString(sessionId, 0);
+					String cipherSuite = sslSession.getCipherSuite();
+					LOGGER.log(LEVEL, "TLS({0},{1},{2})", new Object[] { StringUtil.trunc(sslId, 14), cipherSuite });
+					return new TlsCorrelationContext(id, sslId, cipherSuite);
+				}
+			}
+			LOGGER.log(LEVEL, "TLS*({0})", id);
+			return null;
+		}
+		LOGGER.log(LEVEL, "TCP({0})", id);
+		return new TcpCorrelationContext(id);
+	}
+
+	/**
+	 * Report the correlation context related to the provided channel in future.
+	 * 
+	 * @param channel channel of correlation context
+	 * @param msg message for callback, when context gets established.
+	 */
+	public static void futureCorrelationContext(Channel channel, final RawData msg) {
+		if (null != msg.getMessageCallback()) {
+			SslHandler sslHandler = getSslHandler(channel);
+			if (null != sslHandler) {
+				LOGGER.log(LEVEL, "TLS RAW waiting {0}", toHexString(msg.getBytes(), 16));
+				Future<Channel> handshakeFuture = sslHandler.handshakeFuture();
+				handshakeFuture.addListener(new GenericFutureListener<Future<Channel>>() {
+
+					@Override
+					public void operationComplete(Future<Channel> future) throws Exception {
+						if (future.isSuccess()) {
+							Channel channel = future.getNow();
+							CorrelationContext context = NettyContextUtils.buildCorrelationContext(channel);
+							if (null != context) {
+								msg.onContextEstablished(context);
+							} else {
+								LOGGER.log(Level.WARNING, "Connection {0} still missing valid TLSsession!", channel
+										.id().asShortText());
+							}
+						}
+					}
+				});
+			}
+		}
+	}
+
+	/**
+	 * Byte array to hexadecimal string.
+	 * 
+	 * @param byteArray byte array to be converted to string
+	 * @param max maximum bytes to be converted.
+	 * @return hexadecimal string
+	 */
+	private static String toHexString(byte[] byteArray, int max) {
```

Comment:
OK, but for me this seems to extend the scope of this PR too much.
Separate PR?